certification/internal/authn/keychain.go

```diff
@@ -0,0 +1,81 @@
+package authn
+
+import (
+	"os"
+	"sync"
+
+	"github.com/docker/cli/cli/config"
+	"github.com/docker/cli/cli/config/types"
+	craneauthn "github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+type preflightKeychain struct {
+	mu sync.Mutex
+}
+
+var PreflightKeychain craneauthn.Keychain = &preflightKeychain{}
+
+func (k *preflightKeychain) Resolve(target craneauthn.Resource) (craneauthn.Authenticator, error) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	log.Trace("entering preflight keychain Resolve")
+
+	configFile := viper.GetString("dockerConfig")
+	if configFile == "" {
+		// No file specified. No auth expected
+		return craneauthn.Anonymous, nil
+	}
+
+	if _, err := os.Stat(configFile); err != nil {
+		log.Errorf("Could not find file: %s", configFile)
+		return nil, err
+	}
+
+	r, err := os.Open(configFile)
+	if os.IsNotExist(err) {
+		return craneauthn.Anonymous, nil
+	}
```

Comment:
Is `os.Open` resulting in an `IsNotExist` error cover roughly the same goal as running `os.Stat`? In other words, can we scrap the Stat call?

Otherwise, why do the two error cases for each call return a different outcome? Trying to put together why Stat would find it, but os.Open not finding it means we get Anonymous auth.

Reply:
Let me think about this one a bit more. Your intuition is probably correct.

Reply:
Definitely correct. Fixed.